### ISSUE TITLE
Using -D version=X.Y.Z does not overwrite version usage within conf.py

### ISSUE THREAD

**Issue by shimizukawa** [sphinx-doc/sphinx#1625]
In my conf.py, I am using html_title set to:
  html_title = "Version is " + version

I am then passing in the version variable from the command line via our build system, so
  sphinx-build... -D version=1.0.0 ...

I see that the 'version' variable is updated, but the 'html_title' is not. It is using the default value of 'version' set in the conf.py file.

---
- Bitbucket: https://bitbucket.org/birkenfeld/sphinx/issue/1625
- Originally reported by: Anonymous
- Originally created at: 2014-11-13T06:04:39.984

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2014-11-15 08:26:55+00:00_

This is unfortunately not easily fixed. The conf.py file is just Python code and so it's executed as a whole, and the overrides specified with `-D` are applied afterwards.

**Comment by shimizukawa** [comment]
_From Robert Lehmann on 2014-11-15 09:40:57+00:00_

We _could_ apply the same trick we use for `Config.config_values` and just evaluate `lambda`s at some point.  So the above would have to become `html_title = lambda config: "Version is " + config.version` but that's still better than having nothing at all, right?

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2014-11-15 09:51:23+00:00_

It's a problem with pickling the config, since lambdas don't pickle well.

I was thinking of using a special dict to exec() the conf, prepopulate that with the confoverrides and then ignore assignments to the keys set by confoverrides. It's a bit of a trick, but should have the correct result.

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2014-11-15 10:28:31+00:00_

Hmm, that's probably not going to work, since we need to exec the config in the first place to get the "extensions", otherwise we don't know all possible config values (and can't interpret overrides accordingly).

**Comment by shimizukawa** [comment]
_From Robert Lehmann on 2014-11-17 22:00:49+00:00_

See https://bitbucket.org/birkenfeld/sphinx/pull-request/316 for a proposed change.  Backwards-incompatible though.

**Comment by BuildMonkey-zz** [comment]
Hello-

I filed the original issue on Bitbucket. Any update on when it will get integrated?

Thanks.

**Comment by birkenfeld** [comment]
@BuildMonkey I hope to get this resolved for 1.3. As the original submitter, would you be OK with the solution provided by the pull request linked above? (Having to write `html_title = lambda config: "Version is %s" % config.version`)

**Comment by BuildMonkey-zz** [comment]
I take it that config object is something that is aware of the command line options being passed in, correct? So that the config.version attribute is the value from the CLI. Using lambda is an acceptable solution if this is the case.

Thanks.

**Comment by BuildMonkey-zz** [comment]
I see that there are beta versions of 1.3 released. Is this change in there as well?

Thanks.

**Comment by BuildMonkey-zz** [comment]
I'm still trying to follow up on this issue. Have you integrated the solution into the latest sphinx build or in a prior release of 1.3?

Thanks.

**Comment by tk0miya** [comment]
Note:
Unfortunately, this is not implemented even in Sphinx-1.5.

It emits following warning:
```
WARNING: The config value `html_title' has type `function', expected to ['unicode', 'str'].
```

**Comment by tk0miya** [comment]
Since Sphinx-1.8, `config-inited` event has been added. It can be used to modify config variables dynamically.
For example, following code generates `html_title` using `version` string:
```
def on_config_inited(app, config):
    config.html_title = 'Version is %s' % config.version

def setup(app):
    app.connect('config-inited', on_config_inited)
```


I think this issue has been already resolved.
Closing.
